libc/unistd_fsdir.c

```diff
@@ -0,0 +1,221 @@
+	fclose(fd);
+
+	/* set too long path */
+	memset(toolongpath, 'a', PATH_MAX + 15);
+	toolongpath[PATH_MAX + 15] = '\0';
+}
+
+
+TEST_TEAR_DOWN(unistd_fsdir)
+{
+	TEST_ASSERT_EQUAL_INT(0, remove(filename));
+}
+
+
+TEST(unistd_fsdir, fsdir_getcwd)
+{
+	TEST_ASSERT_NOT_NULL(getcwd(buf, 50));
+	TEST_ASSERT_EQUAL_STRING("/", buf);
+
+	TEST_ASSERT_NULL(getcwd(buf, 0));
+	TEST_ASSERT_EQUAL_INT(EINVAL, errno);
+
+	TEST_ASSERT_NULL(getcwd(buf, 1));
+	TEST_ASSERT_EQUAL_INT(ERANGE, errno);
+}
+
+
+TEST(unistd_fsdir, fsdir_chdir_todir)
+{
+	/* test chdir to some directory */
+	TEST_ASSERT_EQUAL_INT(0, chdir("/dev"));
+	TEST_ASSERT_NOT_NULL(getcwd(buf, 50));
+	TEST_ASSERT_EQUAL_STRING("/dev", buf);
+
+	/* test chdir back to root */
+	TEST_ASSERT_EQUAL_INT(0, chdir("/dev"));
+	TEST_ASSERT_EQUAL_INT(0, chdir("/"));
+	TEST_ASSERT_NOT_NULL(getcwd(buf, 50));
+	TEST_ASSERT_EQUAL_STRING("/", buf);
+}
+
+
+TEST(unistd_fsdir, fsdir_chdir_toolongpath)
+{
+	/* test chdir with too long path */
+	TEST_ASSERT_EQUAL_INT(-1, chdir(toolongpath));
+	TEST_ASSERT_EQUAL_INT(ENAMETOOLONG, errno);
+}
+
+
+TEST(unistd_fsdir, fsdir_chdir_nonexistent)
+{
+	/* test chdir to nonexisting directory */
+	TEST_ASSERT_EQUAL_INT(-1, chdir("/not_existing_directory"));
+	TEST_ASSERT_EQUAL_INT(ENOENT, errno);
+}
+
+
+TEST(unistd_fsdir, fsdir_chdir_emptystring)
+{
+	/* test chdir to empty string */
+	TEST_ASSERT_EQUAL_INT(-1, chdir(""));
+	TEST_ASSERT_EQUAL_INT(ENOENT, errno);
+}
+
+
+TEST(unistd_fsdir, fsdir_chdir_tofile)
+{
+	/* test chdir to file */
+	TEST_ASSERT_EQUAL_INT(-1, chdir(filename));
+	TEST_ASSERT_EQUAL_INT(ENOTDIR, errno);
+}
+
+
+TEST(unistd_fsdir, fsdir_rmdir_empty)
+{
+	/* test removing empty directory */
+	TEST_ASSERT_EQUAL_INT(0, mkdir(directoryname, S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH));
+	TEST_ASSERT_EQUAL_INT(0, rmdir(directoryname));
+}
+
+
+TEST(unistd_fsdir, fsdir_rmdir_nonexistent)
+{
+	/* test rmdir on nonexisting directory*/
+	TEST_ASSERT_EQUAL_INT(-1, rmdir("/not_existing_directory"));
+	TEST_ASSERT_EQUAL_INT(ENOENT, errno);
+}
+
+
+TEST(unistd_fsdir, fsdir_rmdir_toolongpath)
+{
+	/* test rmdir with too long path */
+	TEST_ASSERT_EQUAL_INT(-1, rmdir(toolongpath));
+	TEST_ASSERT_EQUAL_INT(ENAMETOOLONG, errno);
+}
+
+
+TEST(unistd_fsdir, fsdir_rmdir_emptystring)
+{
+	/* test rmdir on empty string */
+	TEST_ASSERT_EQUAL_INT(-1, rmdir(""));
+	TEST_ASSERT_EQUAL_INT(ENOENT, errno);
+}
+
+
+TEST(unistd_fsdir, fsdir_rmdir_file)
+{
+	/* test rmdir on file */
+	TEST_ASSERT_EQUAL_INT(-1, rmdir(filename));
+	TEST_ASSERT_EQUAL_INT(ENOTDIR, errno);
+}
+
+
+TEST(unistd_fsdir, fsdir_rmdir_notempty)
+{
+	/* test removing not empty directory */
+	TEST_ASSERT_EQUAL_INT(0, mkdir(directoryname, S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH));
+	chdir(directoryname);
+	fd = fopen(filename, "w");
+	chdir("/");
+	TEST_ASSERT_EQUAL_INT(-1, rmdir(directoryname));
+	TEST_ASSERT_EQUAL_INT(ENOTEMPTY, errno);
+
+	/* cleanup */
+	chdir(directoryname);
+	remove(filename);
+	chdir("/");
```

Comment:
Missing some asserts to return codes

Reply:
I thought that for cleanup it is not necessary, but it is even clearer when done with asserts. Thanks!